src/webgpu/api/validation/error_scope.spec.ts

```diff
@@ -74,99 +93,187 @@ class F extends Fixture {
+    });
+    for (let i = 0; i < stackDepth; i++) {
+      t.device.pushErrorScope(unmatchedFilters[i % unmatchedFilters.length]);
+    }
 
-g.test('push,popping_nested_error_scopes_must_be_balanced').fn(async t => {
-  {
-    const promise = t.device.popErrorScope();
-    t.shouldReject('OperationError', promise);
-  }
+    // Current scope should catch the error immediately.
+    t.device.pushErrorScope(errorFilter);
+    t.generateError(errorFilter);
+    const error = await t.device.popErrorScope();
+    t.expect(t.isInstanceOfError(errorFilter, error));
+
+    // Remaining scopes shouldn't catch anything.
+    const promises = [];
+    for (let i = 0; i < stackDepth; i++) {
+      promises.push(t.device.popErrorScope());
+    }
+    const errors = await Promise.all(promises);
+    t.expect(errors.every(e => e === null));
+  });
 
-  const promises = [];
-  for (let i = 0; i < 1000; i++) {
-    t.device.pushErrorScope('validation');
-  }
-  for (let i = 0; i < 1000; i++) {
-    promises.push(t.device.popErrorScope());
-  }
-  const errors = await Promise.all(promises);
-  t.expect(errors.every(e => e === null));
+g.test('balanced_siblings')
+  .desc(
+    `
+Tests that sibling error scopes need to be balanced.
+
+- Different error types as the current scope
+- Different number of sibling errors
+    `
+  )
+  .params(u =>
+    u.combine('errorFilter', kErrorScopeFilters).combine('numErrors', [1, 10, 100, 1000, 100000])
+  )
+  .fn(async t => {
+    const { errorFilter, numErrors } = t.params;
+
+    const promises = [];
+    for (let i = 0; i < numErrors; i++) {
+      t.device.pushErrorScope(errorFilter);
+      promises.push(t.device.popErrorScope());
+    }
+    const errors = await Promise.all(promises);
+    t.expect(errors.every(e => e === null));
+
+    {
+      // Trying to pop an additional non-exisiting scope should reject.
+      const promise = t.device.popErrorScope();
+      t.shouldReject('OperationError', promise);
+    }
+  });
 
-  {
-    const promise = t.device.popErrorScope();
-    t.shouldReject('OperationError', promise);
-  }
-});
+g.test('balanced_nesting')
+  .desc(
+    `
+Tests that nested error scopes need to be balanced.
+
+- Different error types as the current scope
+- Different number of nested errors
+    `
+  )
+  .params(u =>
+    u.combine('errorFilter', kErrorScopeFilters).combine('numErrors', [1, 10, 100, 1000, 100000])
+  )
+  .fn(async t => {
+    const { errorFilter, numErrors } = t.params;
+
+    for (let i = 0; i < numErrors; i++) {
+      t.device.pushErrorScope(errorFilter);
+    }
+
+    const promises = [];
+    for (let i = 0; i < numErrors; i++) {
+      promises.push(t.device.popErrorScope());
+    }
+    const errors = await Promise.all(promises);
+    t.expect(errors.every(e => e === null));
+
+    {
+      // Trying to pop an additional non-exisiting scope should reject.
+      const promise = t.device.popErrorScope();
+      t.shouldReject('OperationError', promise);
+    }
```

Comment:
Same although maybe leave one as-is just for variety.

Reply:
See comment above.